### PR TITLE
feature/AkD-327: Plugin to check recently active IAM user

### PR DESCRIPTION
**Pull request by UmairAjmal**

### CODE REVIEW

File: plugins/aws/iam/iamUserInUse.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
+var helpers = require('../../../helpers/aws');
+var async = require('async');
+
+module.exports = {
+    title: 'IAM User Account In Use',
+    category: 'IAM',
+    description: 'Ensures the IAM user account is being actively used',
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamUserInUse.js:7]:
```suggestion
    description: 'Ensure that IAM user accounts are being actively used.',
```

File: plugins/aws/iam/iamUserInUse.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    compliance: { // Need comments for these statements
+        hipaa: 'HIPAA requires strong auditing controls surrounding actions ' +
+                'taken in the environment. The IAM user user lacks these controls ' +
+                'since it is not tied to a specific user. The IAM user account ' +
+                'should not be used.',
+        pci: 'PCI requires that cardholder data can only be accessed by those with ' +
+             'a legitimate business need. Restricting IAM user access prevents access ' +
+             'to these environments from users who may not be identified.',
+        cis1: '1.1 Avoid the use of the IAM user account'
+    },
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamUserInUse.js:21]:
```suggestion
```

File: plugins/aws/iam/iamUserInUse.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            regex: '^[1-9]{1}[0-9]{0,3}$',
+            default: 15
+        }
+    },
+
+    run: function(cache, settings, callback) {
+        this._run(cache, settings, callback, new Date());
+    },
+
+    _run: function(cache, settings, callback, now) {
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamUserInUse.js:35]: Change this to be consistent with newer plugins

File: plugins/aws/iam/iamUserInUse.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+        var found = false;
+
+        async.forEach(generateCredentialReport.data, (user) => {
+            if (user && user.user !== '<root_account>') {
+                found = true;
+
+                var accessDates = [];
+
+                if (user.password_last_used && user.password_last_used !== null && user.password_last_used !== 'no_information') {
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamUserInUse.js:66]:
```suggestion
                if (user.password_last_used && user.password_last_used !== 'no_information') {
```

File: plugins/aws/iam/iamUserInUse.js [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+                    accessDates.push(user.password_last_used);
+                }
+
+                if (user.access_key_1_last_used_date && user.access_key_1_last_used_date !== null) {
+                    accessDates.push(user.access_key_1_last_used_date);
+                }
+
+                if (user.access_key_2_last_used_date && user.access_key_2_last_used_date !== null) {
+                    accessDates.push(user.access_key_2_last_used_date);
+                }
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamUserInUse.js:76]:
```suggestion
                if (user.access_key_1_last_used_date ) accessDates.push(user.access_key_1_last_used_date);

                if (user.access_key_2_last_used_date) accessDates.push(user.access_key_2_last_used_date);
```

File: plugins/aws/iam/iamUserInUse.js [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+                if (!accessDates.length) {
+                    helpers.addResult(results, 0, 'IAM user has not been used', 'global', user.arn);
+                } else {
+                    var dateToCompare = helpers.mostRecentDate(accessDates);
+                    console.log(accessDates, dateToCompare, now, helpers.daysBetween(dateToCompare, now));
+
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamUserInUse.js:83]:
```suggestion
                    var dateToCompare = helpers.mostRecentDate(accessDates);

```

File: plugins/aws/iam/iamUserInUse.js [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+            }
+        });
+
+        if (!found) {
+            helpers.addResult(results, 3, 'Unable to query IAM users');
+        }
+
+        callback(null, results, source);
+
+    }
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamUserInUse.js:99]:
```suggestion
        callback(null, results, source);
    }
```

File: plugins/aws/iam/iamUserInUse.spec.js [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+            const callback = (err, results) => {
+                console.log(results)
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(2);
+                expect(results[0].message).to.equal('IAM user was last used 1 days ago');
+                done();
+            };
+
+            iamUserInUse._run(cache, settings, callback, new Date('2021-07-06T15:31:34+00:00'));
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamUserInUse.spec.js:57]: Make this dynamic

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: plugins/aws/iam/iamUserInUse.js [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+                    helpers.addResult(results, resultCode,
+                        'IAM user was last used ' + helpers.daysBetween(dateToCompare, currentDate) + ' days ago',
+                        'global', user.arn, custom);
+                }
+            }
+        });
+
+        if (!found) {
+            helpers.addResult(results, 3, 'Unable to query IAM users');
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamUserInUse.js:76]:
```suggestion
            helpers.addResult(results, 0, 'No IAM users found', 'global');
```

File: exports.js [hunk 10]
```diff
@@ -235,6 +235,7 @@ module.exports = {
         'usersPasswordAndKeys'          : require(__dirname + '/plugins/aws/iam/usersPasswordAndKeys.js'),
         'usersPasswordLastUsed'         : require(__dirname + '/plugins/aws/iam/usersPasswordLastUsed.js'),
         'canaryKeysUsed'                : require(__dirname + '/plugins/aws/iam/canaryKeysUsed.js'),
+        'iamUserInUse'                : require(__dirname + '/plugins/aws/iam/iamUserInUse.js'),
```
Comment (AkhtarAmir) [inline exports.js:238]:
```suggestion
        'iamUserInUse'                  : require(__dirname + '/plugins/aws/iam/iamUserInUse.js'),
```

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: plugins/aws/iam/iamUserInUse.js [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    category: 'IAM',
+    description: 'Ensure that IAM user accounts are being actively used.',
+    more_info: 'IAM users, roles, and groups should be used for day-to-day account management.',
+    link: 'https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users_create.html',
+    recommended_action: 'Delete IAM user accounts which are not being actively used.',
+    apis: ['IAM:generateCredentialReport'],
+    settings: {
+        iam_user_account_in_use_days: {
+            name: 'IAM User Account In Use Days',
+            description: 'Return a failing result when the IAM user account has been used within this many days',
```
Comment (giorod3) [inline plugins/aws/iam/iamUserInUse.js:15]:
```suggestion
            description: 'Return a failing result when the IAM user account has not been used within this many days',
```

**Review by giorod3 — CHANGES_REQUESTED** [review]
review the logic, this should be checking to see if a user is use. it should fail if a user isn't being actively used to prevent lingering users

File: plugins/aws/iam/iamUserInUse.js [hunk 12]
```diff
@@ -0,0 +1,79 @@
+var helpers = require('../../../helpers/aws');
+var async = require('async');
+
+module.exports = {
+    title: 'IAM User Account In Use',
+    category: 'IAM',
+    description: 'Ensure that IAM user accounts are not being actively used.',
+    more_info: 'IAM users, roles, and groups should not be used for day-to-day account management.',
+    link: 'https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users_create.html',
+    recommended_action: 'Delete IAM user accounts which are not being actively used.',
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamUserInUse.js:10]:
```suggestion
    recommended_action: 'Delete IAM user accounts which are being actively used.',
```

File: plugins/aws/iam/iamUserInUse.js [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+                    helpers.addResult(results, 0, 'IAM user has not been used', 'global', user.arn);
+                } else {
+                    var currentDate = new Date();
+                    var loginDate = new Date(helpers.mostRecentDate(accessDates));
+                    var resultCode = (helpers.daysBetween(loginDate, currentDate) < config.iam_user_account_in_use_days) ? 2: 0;
+
+                    helpers.addResult(results, resultCode,
+                        'IAM user was last used ' + helpers.daysBetween(loginDate, currentDate) + ' days ago',
+                        'global', user.arn, custom);
+                }
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamUserInUse.js:69]:
```suggestion
                    var loginDate = new Date(helpers.mostRecentDate(accessDates));
                    var difference = helpers.daysBetween(loginDate, currentDate);
                    var resultCode = (difference < config.iam_user_account_in_use_days) ? 2: 0;

                    helpers.addResult(results, resultCode,
                        'IAM user was last used ' + difference + ' days ago',
                        'global', user.arn, custom);
                }
```

File: plugins/aws/iam/iamUserInUse.js [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+            if (user && user.user !== '<root_account>') {
+                found = true;
+
+                var accessDates = [];
+
+                if (user.password_last_used && user.password_last_used !== 'no_information') {
+                    accessDates.push(user.password_last_used);
+                }
+
+                if (user.access_key_1_last_used_date) accessDates.push(user.access_key_1_last_used_date);
```
Comment (AkhtarAmir) [inline plugins/aws/iam/iamUserInUse.js:55]: Check if (user.access_key_1_last_used_date != 'N/A')

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

**Review by AkhtarAmir — APPROVED** [review]

**Review by giorod3 — APPROVED** [review]

**Review by giorod3 — APPROVED** [review]

**Pull request merged**